pkg/ipcache/ipcache.go

```diff
@@ -680,15 +684,18 @@ func (ipc *IPCache) deleteLocked(ip string, source source.Source) (namedPortsCha
 
 // GetNamedPorts returns a copy of the named ports map. May return nil.
 func (ipc *IPCache) GetNamedPorts() (npm types.NamedPortMultiMap) {
-	ipc.mutex.Lock()
-	if !ipc.needNamedPorts {
-		ipc.needNamedPorts = true
-		ipc.updateNamedPorts()
+	ipc.mutex.RLock()
+	defer ipc.mutex.RUnlock()
```

Comment:
What's the purpose of the Read lock here?

Reply:
Oh, it looks like this is because `updateNamedPortsRLocked() needs to access `ipc.ipToK8sMetadata` to read & prepare the new map to atomically update. We could probably shift that locking into the child function since I'm not seeing anything inside this function that relies on the readlock usage.

Reply:
^ But if we moved it into the child function, we'd also need to provide a "don't lock" variant for `deleteLocked()` and friends. So probably this is reasonable enough to leave here, it's just that putting the locking at the start of the function implies additional locking protections than just the need for stable read access for the child call. 🤷 it works, I just want to make sure we don't make assumptions about multiple threads inside this function because that *can* be the case. Based on the usage of atomics etc. here I think the behaviour is correct.

Reply:
Once we land the [ipcache refactor](https://github.com/cilium/cilium/issues/21142), hopefully this will clean up this code path.

Reply:
@joestringer @christarazi We just determined that this RLock here will still cause a deadlock:

We have `EndpointRLocked -> IPCacheRLocked` (for `GetNamedPorts`) and `IPCacheWLocked -> EndpointWLocked` (for `InjectLabels`), which will deadlock if the two paths are taken by to go routines in parallel.

I will investigate if it's possible to refactor the `GetNamedPorts` to not hold the RLock here. Otherwise, I think we need to introduce the additional [IPCache update lock](https://github.com/cilium/cilium/compare/v1.12...gandro:cilium:pr/gandro/v1.12-ipcache-deadlock-attempt-v5?expand=1) that allows for the IPCache RLock to be taken during InjectLabels.

Reply:
I've pushed a commit that just always computes the `NamedPorts` eagerly, regardless if it's used by a policy or not. I'm not sure how expensive that will be: https://github.com/cilium/cilium/pull/24672/commits/3a5e8d9fa06a31aa3929f437fde9d5d33d77327c